tensorflow/python/ops/init_ops.py

```diff
@@ -340,3 +341,45 @@ def __call__(self, shape, dtype=dtypes.float32, partition_info=None):
     """Generate a tensor used to initialize a variable."""
     return random_ops._random_walk(shape, self._nonlinearity, dtype,
                                    seed=self._seed)
+
+
+def orthogonal_initializer(gain=1.0, dtype=dtypes.float32, seed=None):
+  """Returns an initializer that generates an orthogonal rotation matrix.
+
+  Initialize a tensor using an orthogonal matrix.
+
+  Args:
+    gain: multiplicative factor to apply to the orthogonal matrix
+    dtype: The type of the output.
+    seed: A Python integer. Used to create random seeds. See
+      [`set_random_seed`](../../api_docs/python/constant_op.md#set_random_seed)
+      for behavior.
+
+  Returns:
+    An initializer that generates orthogonal tensors
+
+  Raises:
+    ValueError: if `dtype` is not a floating point type.
+  """
+  def _initializer(shape, dtype=_assert_float_dtype(dtype), partition_info=None):
+    # Flatten the input shape with the last dimension remaining its original shape so it works for conv2d
```

Comment:
I think we need to clearly document what this shape means in the docstring: what is supported, and what is orthogonal as the output. 

Since even conv has multiple data formats, NHWC and NCHW, it is a good idea to separate them clearly. It is a good idea to keep this op has simple as possible, and requires the shape to be square matrices.